### PR TITLE
Issue 6251: NE: Added destroy function to DM_DECLARE_COMPONENT_TYPE

### LINKED ISSUE #6251: Add destroy function to DM_DECLARE_COMPONENT_TYPE
In order to be able to better handle life cycle control of the component extensions (e.g. free context memory), we need to add a destroy function as well.

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: engine/gameobject/src/dmsdk/gameobject/component.h [hunk 1]
```diff
@@ -439,6 +439,14 @@ namespace dmGameObject
     typedef void (*ComponentIterProperties)(dmGameObject::SceneNodePropertyIterator* pit, dmGameObject::SceneNode* node);
 
 
+    /*#
+     * Get the component type index. Used for with e.g. dmGameObject::GetWorld()/GetContext()
+     * @name ComponentTypeGetTypeIndex
+     * @param type [type: ComponentType*] the type
+     * @return type_index [type: uint32_t] The type index.
+     */
+    uint32_t ComponentTypeGetTypeIndex(const ComponentType* type);
```
Comment (JCash) [inline engine/gameobject/src/dmsdk/gameobject/component.h:448]: Good to have for script functions for components

File: engine/gameobject/src/dmsdk/gameobject/component.h [hunk 2]
```diff
@@ -579,10 +587,18 @@ namespace dmGameObject
...
      */
     void ComponentTypeSetContext(ComponentType* type, void* context);
 
+    /*# get the component type global context
+     * get the component type global context
+     * @name ComponentTypeGetContext
+     * @param type [type: ComponentType*] the type
+     * @return context [type: void*] component type global context
+     */
+    void* ComponentTypeGetContext(const ComponentType* type);
```
Comment (JCash) [inline engine/gameobject/src/dmsdk/gameobject/component.h:600]: Used to get the context when shutting down the engine.

File: engine/gameobject/src/dmsdk/gameobject/component.h [hunk 3]
```diff
@@ -666,24 +683,27 @@ namespace dmGameObject
...
-
-    /*#
+    /*# register a new component type
+     * @param symbol [type: cpp_symbol_name] The unique C++ symbol name
+     * @param name [type: const char*] name of the component type (i.e. the resource suffix)
+     * @param create_fn [type: dmGameObject::Result (*fn)(const ComponentTypeCreateCtx* ctx, ComponentType* type)] The type configuration function. May not be 0.
+     * @param destroy_fn [type: dmGameObject::Result (*fn)(const ComponentTypeCreateCtx* ctx, ComponentType* type)] The type destruction function. May be 0.
      */
-    #define DM_DECLARE_COMPONENT_TYPE(symbol, name, type_create_fn) \
+    #define DM_DECLARE_COMPONENT_TYPE(symbol, name, type_create_fn, type_destroy_fn) \
```
Comment (JCash) [inline engine/gameobject/src/dmsdk/gameobject/component.h:704]: A breaking change!

Only used in 2 extensions (spine+rive), so should create new fix-branches for those.

File: engine/gameobject/src/gameobject/gameobject.cpp [hunk 4]
```diff
@@ -471,18 +471,47 @@ namespace dmGameObject
...
+        return 0xFFFFFFFF;
+    }
+
+    void* GetWorld(HCollection hcollection, uint32_t component_type_index)
+    {
+        Register* regist = GetRegister(hcollection);
+        if (component_type_index < regist->m_ComponentTypeCount)
         {
-            return hcollection->m_Collection->m_ComponentWorlds[component_index];
+            return hcollection->m_Collection->m_ComponentWorlds[component_type_index];
```
Comment (JCash) [inline engine/gameobject/src/gameobject/gameobject.cpp:493]: A better name imho. Not to be confused with the component indices used in the collection.

File: engine/gameobject/src/gameobject/test/component/test_gameobject_component.cpp [hunk 5]
```diff
@@ -615,6 +615,106 @@ TEST_F(ComponentTest, FinalCallsFinal)
...
+    ComponentTypeSetContext(type, g_ComponentApiTestContext.m_CreateContext);
+    return dmGameObject::RESULT_OK;
+}
+
+static dmGameObject::Result ComponentTypeTest_Destroy(const dmGameObject::ComponentTypeCreateCtx* ctx, dmGameObject::ComponentType* type)
+{
+    g_ComponentApiTestContext.m_Destroyed = 1;
+    g_ComponentApiTestContext.m_DestroyContext = ComponentTypeGetContext(type);
+
+    ComponentTypeSetContext(type, g_ComponentApiTestContext.m_CreateContext);
```
Comment (JCash) [inline engine/gameobject/src/gameobject/test/component/test_gameobject_component.cpp:654]: Normally, you'd delete the context here (the point of this PR), but we keep it in order to check that the context is stored/retrieved properly.

File: engine/gameobject/src/dmsdk/gameobject/component.h [hunk 6]
```diff
@@ -632,7 +648,7 @@ namespace dmGameObject
      * @member m_Factory [type: dmResource::HFactory] The resource factory
      * @member m_Register [type: dmGameObject::HRegister] The game object registry
      * @member m_Script [type: dmScript::HContext] The shared script context
-     * @member m_Contextx [type: dmHashTable64<void*>] Mappings between names and contextx
+     * @member m_Contexts [type: dmHashTable64<void*>] Mappings between names and contextx
```
Comment (britzl) [inline engine/gameobject/src/dmsdk/gameobject/component.h:651]: contextx -> contexts

**Review by britzl — APPROVED** [review]
Looks good

**Pull request merged**